overlord/snapstate/reboot.go

```diff
@@ -123,6 +176,216 @@ func bootBaseSnapType(byTypeTss map[snap.Type]*state.TaskSet) snap.Type {
 	return snap.TypeOS
 }
 
+// configureSingleRebootForSplitTaskSets sets up restart boundaries for task-sets that have been
+// split into two. It uses the first part of the task-set (i.e the one containing the reboot edges)
+// to set a restart boundary along the 'Do' path for the last task-set, and one along the 'Undo' path
+// for the first task-set, to introduce only one reboot along each direction.
+func configureSingleRebootForSplitTaskSets(byTypeTss, beforeTss map[snap.Type]*state.TaskSet) {
+	// Set reboot boundary along the do path for the last essential
+	for i := len(essentialSnapsRestartOrder) - 1; i >= 0; i-- {
+		o := essentialSnapsRestartOrder[i]
+		if beforeTss[o] == nil {
+			continue
+		}
+
+		// refer to the original task-set that has edges
+		if linkSnap := byTypeTss[o].MaybeEdge(MaybeRebootEdge); linkSnap != nil {
+			restart.MarkTaskAsRestartBoundary(linkSnap, restart.RestartBoundaryDirectionDo)
+		}
+		break
+	}
+
+	// Set reboot boundary along the undo path for the first essential
+	for _, o := range essentialSnapsRestartOrder {
+		if beforeTss[o] == nil {
+			continue
+		}
+
+		// just for consistency lets refer to original task-set
+		if unlinkSnap := findUnlinkTask(byTypeTss[o]); unlinkSnap != nil {
+			restart.MarkTaskAsRestartBoundary(unlinkSnap, restart.RestartBoundaryDirectionUndo)
+		}
+		break
+	}
+}
+
+// configureSnapToWaitForBaseIfPresent sets up dependency on the base of a snap, if the base is
+// also being updated.
+func configureSnapToWaitForBaseIfPresent(snapTs *state.TaskSet, bases map[string]*state.TaskSet) error {
+	snapsup := maybeTaskSetSnapSetup(snapTs)
+	if snapsup == nil {
+		return fmt.Errorf("internal error: failed to get the SnapSetup instance from snap task-set")
+	}
+
+	if baseTs := bases[snapsup.Base]; baseTs != nil {
+		snapTs.WaitAll(baseTs)
```

Comment:
I understand it's hard be we should try to avoid using WaitAll across lanes, it makes the result task graph really hard to follow